src/niri.rs

```diff
@@ -3730,45 +3745,78 @@ impl Niri {
 
     pub fn render_for_screencopy(
         &mut self,
-        backend: &mut Backend,
+        renderer: &mut GlesRenderer,
+        is_winit_backend: bool,
         screencopy: Screencopy,
     ) -> anyhow::Result<()> {
+        let _span = tracy_client::span!("Niri::render_for_screencopy");
+
         let output = screencopy.output().clone();
         ensure!(self.output_state.contains_key(&output), "output is missing");
 
         self.layout.update_render_elements(&output);
 
-        backend
-            .with_primary_renderer(move |renderer| {
-                let elements = self
-                    .render(
-                        renderer,
-                        &output,
-                        screencopy.overlay_cursor(),
-                        RenderTarget::ScreenCapture,
-                    )
-                    .into_iter()
-                    .rev();
-
-                let region_loc = screencopy.region_loc();
-                let elements = elements.map(|element| {
-                    RelocateRenderElement::from_element(
-                        element,
-                        region_loc.upscale(-1),
-                        Relocate::Relative,
-                    )
-                });
+        let elements = self.render(
+            renderer,
+            &output,
+            screencopy.overlay_cursor(),
+            RenderTarget::ScreenCapture,
+        );
 
-                let scale = output.current_scale().fractional_scale().into();
-                let transform = output.current_transform();
-                render_to_shm(renderer, screencopy.buffer(), scale, transform, elements)
-                    .context("error rendering to screencopy shm buffer: {err:?}")?;
+        let damage_tracker = self.screencopy_state.damage_tracker_or_insert(&output);
+        let damages = damage_tracker
+            .output_damage_tracker()
+            .damage_output(1, &elements)
+            .context("output has no active mode")?
+            .0;
 
-                screencopy.submit(false);
+        if screencopy.with_damage() && damages.is_none() {
+            // No damage, wait until there is some when redrawing the output.
+            *damage_tracker.awaiting_copy() = Some(screencopy);
+            return Ok(());
+        }
 
-                Ok(())
-            })
-            .context("primary renderer is missing")?
+        let region_loc = screencopy.region_loc();
+        let elements = elements.into_iter().rev().map(|element| {
+            RelocateRenderElement::from_element(element, region_loc.upscale(-1), Relocate::Relative)
+        });
```

Comment:
This should happen before the damage tracker because this will affect the damage. (If you screencopy a small rectangle of the output, all damage should be computed only in this rectangle and relative to it.)

Reply:
Indeed since the x and y are both uint's. I will clip and transform relative to the region_loc then.